### DISCUSSION TITLE
Setting request header or prefer c.get()?

### DISCUSSION THREAD

**Discussion by exoup** [honojs/hono#2847] [Q&A] [answered]
Howdy!

I'm writing a middleware to pass some necessary information in requests that my application receives.

Is there a known method to set a custom request header for a HonoRequest or is it preferred not to modify the request header and instead just pass required data via the Context getter and setter?

Can I modify c.req.raw.header somehow?

**Comment by 6km** [comment, marked as answer]
Hi @exoup 

It's possible to pass data using the following techniques (not just these)

### Setting a custom `request` header
It can be done using the `c.req.raw.headers.set()` method but it should be before calling `next()`.

Please notice that when you set a `request` header from the server side, it will not be passed to the client side. This technique is only used to pass data to the route handler.

```ts
app.use(async (c, next) => {
  c.req.raw.headers.set('x-username', '6km')
  await next()
})

app.get('/', (c) => {
  const passedData = c.req.header('x-username')
  // ...
})
```

---

### Setting a custom `response` header

This technique is used to pass data to the server through response headers.

To set a **response** header, it's simple, just use `c.header(name, value)`.

```ts
app.use(async (c, next) => {
  await next()
  c.header('x-username', '6km')
})
```

![image](https://github.com/honojs/hono/assets/62352949/dc08d77b-1fc8-4c68-8cfd-753404d2ad9a)

---

### Passing data through `Context`

Set the value specified by the key with `set` and use it later with `get`.
This can be used to pass methods and additional data.

```ts
type Variables = {
  username: string
}

const app = new Hono<{Variables: Variables}>()

app.use(async (c, next) => {
  c.set('username', '6km')
  await next()
})

app.get('/', (c) => {
  const passedData  = c.get('username')
  // ...
})
```

**Comment by exoup** [comment]
@6km Appreciate the thorough response.

I was initially using `c.req.raw.headers.set()` to set my headers but kept running into a `TypeError: Headers are immutable` error.

Despite not using upgradeWebSocket anywhere, I suspect I'm running into this issue #2535.
I'll just stick with the Context object.

Thanks!

**Reply by 6km** [reply]
You're welcome! By the way, which version of Hono are you using? Try the latest one.

**Reply by exoup** [reply]
> You're welcome! By the way, which version of Hono are you using? Try the latest one.

I was using 4.3.11 but the issue also exists in 4.4.0, but it's all right.

**Reply by 6km** [reply]
> I was using 4.3.11 but the issue also exists in 4.4.0, but it's all right.

Oh, I hope it gets fixed asap!

**Reply by yusukebe** [reply]
Hi. Regardless of whether you are using `upgradeWebSocket` or not, `c.req.raw.header` - a `Headers` object can be immutable. So, please use `c.header()`!
